ext/SpeedyWeatherEnzymeExt.jl

```diff
@@ -0,0 +1,115 @@
+module SpeedyWeatherEnzymeExt
+
+using SpeedyWeather
+using Enzyme
+import .EnzymeRules: reverse, augmented_primal
+using .EnzymeRules
+
+# import all functions for which we define rules
+import SpeedyWeather.SpeedyTransforms: _fourier!
+
+# Rules for SpeedyTransforms
+
+# _fourier! 
+
+# Computes the scale for the adjoint/pullback of all discrete Fourier transforms. 
+function adjoint_scale(S::SpectralTransform)
+    (; nlat_half, nlons, rfft_plans) = S
+    nfreqs = [rfft_plan.osz[1] for rfft_plan in rfft_plans] # TODO: This works with FFTW, but does it with CUFFT as well?
+
+    scale = zeros(Int, maximum(nfreqs), nlat_half) 
+
+    for i=1:nlat_half
+        scale[1:nfreqs[i],i] = rfft_adjoint_scale(nfreqs[i], nlons[i])
+    end 
+
+    # TODO: transfer array to GPU in case we are on GPU
+    return reshape(scale, maximum(nfreqs), 1, nlat_half) # the scratch memory is (Freq x lvl x lat), so we insert 
+                                                         # an additional dimension here for easier matrix multiply
+end 
+
+# Computes the scale for the adjoint/pullback of a real discrete fourier transform.
+function rfft_adjoint_scale(n_freq::Int, n_real::Int)
+    if iseven(n_real)
+        return [1; [2 for i=2:(n_freq-1)]; 1]
+    else 
+        return [1; [2 for i=2:n_freq]]
+    end 
+end 
+
+### Custom rule for _fourier!(f_north, f_north, grid, S)
+function augmented_primal(config::EnzymeRules.RevConfigWidth{1}, func::Const{typeof(_fourier!)}, ::Type{<:Const}, 
+    f_north::Duplicated, f_south::Duplicated, grids::Duplicated{<:AbstractGridArray}, S::Union{Const, MixedDuplicated}) 
+
+    func.val(f_north.val, f_south.val, grids.val, S.val) # forward pass
+
+    # save grids in tape if grids will be overwritten
+    if overwritten(config)[4] # TODO: Not sure this is really necessary because grids won't ever get overwritten by this _fourier!
+        tape = copy(grids.val)
+    else
+        tape = nothing
+    end
+
+    return AugmentedReturn(nothing, nothing, tape) # because the function actually returns nothing
+
+end 
+
+function reverse(config::EnzymeRules.RevConfigWidth{1}, func::Const{typeof(_fourier!)}, ::Type{<:Const}, tape,
+    f_north::Duplicated, f_south::Duplicated, grids::Duplicated{<:AbstractGridArray}, S::Union{Const, MixedDuplicated})
```

Comment:
So `::Any` becomes `::Duplicated` and `::T` becomes `::Duplicated{<:T}`? And that's how to specify the method of a function?

Reply:
I didn't find this documented anywhere, but this is how I think it works.